File: packages/convex/convex/lib/affinity.ts
import type { RecommendationEventAction, RecommendationInteraction } from "@pantry/types";

/**
 * Turning the interaction log into a taste signal (BL-0005 increment 2).
 *
 * This is a PURE function of (events, now) and lives away from any Convex
 * handler so it can be table-tested. That is not just tidiness: the rules below
 * are the entire learned half of the recommender, and every one of them is a
 * judgement about how to read a user's behaviour rather than a mechanism.
 *
 * Three decisions are load-bearing:
 *
 *  - **Derived at request time, never stored as a score.** There is no
 *    per-user per-ingredient table. A stored score is a second source of truth
 *    about the same events, and the day it drifts from them there is no way to
 *    tell which one is wrong. If this fold ever becomes expensive it becomes a
 *    CACHE of these events — a local change with an obvious invalidation — not
 *    a new authority.
 *  - **An empty fold returns an EMPTY map, and the ranker treats that as
 *    UNAVAILABLE rather than as a row of zeroes.** A user who has done nothing
 *    has told us nothing. Sending `{}` and having the ranker score every
 *    candidate at zero on the heaviest-weighted feature would punish precisely
 *    the users who have not used the product yet. See
 *    `internal/recommend/affinity.go`.
 *  - **Saturating, not relative, normalization.** Weights are squashed with
 *    tanh against a fixed scale rather than divided by the largest weight
 *    present. Relative normalization would make the top ingredient score ±1 for
 *    everyone forever — a user with one click and a user with a year of history
 *    would look equally certain — and would rescale every other ingredient every
 *    time a new event arrived.
 */

/**
 * What each kind of interaction says about taste.
 *
 * `shown` is deliberately ZERO. An impression is not an opinion: the user did
 * not choose to be shown the card, and counting it would let the recommender
 * teach itself its own past decisions. Impressions are still recorded, and they
 * are read by the discovery ranker's `novelty` — "you have seen this six times"
 * is a fact about the UI, and belongs where facts about the UI belong.
 *
 * `cooked` outweighs `accepted` because planning a meal is an intention and
 * cooking it is a completed act; the gap between the two is exactly the signal
 * we would lose by treating the plan as the outcome.
 *
 * `dismissed` is heavier than `accepted` on purpose. Rejection is the rarer and
 * more deliberate action — a user scrolls past a hundred things they are mildly
 * uninterested in and dismisses the one they actively do not want.
 */
const ACTION_WEIGHT: Record<RecommendationEventAction, number> = {
  cooked: 1,
  accepted: 0.6,
  dismissed: -0.8,
  shown: 0,
};

/**
 * How long a signal takes to lose half its strength.
 *
 * Tastes change, and a month is roughly the horizon over which "I have been
 * cooking a lot of Thai food" stops being true without anyone saying so. Long
 * enough that a fortnight's holiday does not wipe a user's profile; short enough
 * that last spring's phase does not outvote this week's cooking.
 */
const HALF_LIFE_MS = 30 * 24 * 60 * 60 * 1000;

/**
 * How much accumulated evidence approaches a full ±1.
 *
 * At this scale a single accepted suggestion moves an ingredient to ~0.29 and a
 * single cook to ~0.46 — present, but nothing like the ±1 a user gets for typing
 * the ingredient into their settings. Three cooks reach ~0.9. That ordering is
 * the point: what someone tells us outranks what we inferred, until we have
 * inferred it many times over.
 */
const AFFINITY_SCALE = 2;

/**
 * Below this an ingredient is noise, and shipping it would just cost payload for
 * a weight that cannot change an ordering.
 */
const AFFINITY_FLOOR = 0.05;

/**
 * How many ingredient weights travel in the request. The ranker looks each
 * candidate's ingredients up in this map, so the tail contributes almost
 * nothing while costing bytes on every single recommendation call.
 */
const MAX_AFFINITIES = 50;

/** One row of the event log, reduced to what the fold reads. */
export interface FoldableEvent {
  recipeId: string;
  action: RecommendationEventAction;
  /**
   * The recipe's canonical ingredients AS THEY WERE when the event happened.
   *
   * Denormalized onto the event for the same reason `nutritionLog.snapshot` is:
   * an event is a historical fact, and re-deriving it from a recipe that has
   * since been edited or deleted would silently rewrite what the user did.
   * Convex also cannot look ingredients up on its own — recipe bodies live in
   * recipe-service, and a mutation cannot fetch.
   *
   * Absent on events recorded without them. Such an event still counts toward
   * `interactions` (which is per recipe) and contributes nothing to
   * `affinities` (which is per ingredient) — a partial record, honestly read.
   */
  canonicalItems?: string[];
  createdAt: number;
}

export interface DerivedSignal {
  /** canonicalItem → [-1, 1]. Empty means "no signal", not "all neutral". */
  affinities: Record<string, number>;
  /** recipeId → recent interaction counts, for the discovery ranker's novelty. */
  interactions: Record<string, RecommendationInteraction>;
}

/** Half-life decay: 1 at the moment of the event, 0.5 one half-life later. */
function recencyWeight(createdAt: number, now: number): number {
  const age = now - createdAt;
  // A clock skew that puts an event in the future is not evidence that it
  // matters more than the present. Clamp rather than amplify.
  if (age <= 0) return 1;
  return 0.5 ** (age / HALF_LIFE_MS);
}

/**
 * Fold a recent window of events into the derived signal the ranker reads.
 *
 * The caller decides the window; this function only decays what it is given.
 * That split keeps the recency policy where the database query is and keeps this
 * function a pure function of its arguments.
 */
export function deriveSignal(events: FoldableEvent[], now: number): DerivedSignal {
  const raw: Record<string, number> = {};
  const interactions: Record<string, RecommendationInteraction> = {};

  for (const event of events) {
    interactions[event.recipeId] ??= { shown: 0, accepted: 0, dismissed: 0, cooked: 0 };
    interactions[event.recipeId][event.action] += 1;

    const weight = ACTION_WEIGHT[event.action];
    // `shown` weighs nothing, so it never reaches the ingredient map at all —
    // it has already done its job in `interactions` above.
    if (weight === 0 || event.canonicalItems === undefined) continue;

    const decayed = weight * recencyWeight(event.createdAt, now);
    // De-duplicated per event: a recipe listing garlic twice is one opinion
    // about garlic, not two.
    for (const item of new Set(event.canonicalItems)) {
      if (item === "") continue;
      raw[item] = (raw[item] ?? 0) + decayed;
    }
  }

  const affinities: Record<string, number> = {};
  const scored = Object.entries(raw)
    .map(([item, sum]) => [item, Math.tanh(sum / AFFINITY_SCALE)] as const)
    .filter(([, value]) => Math.abs(value) >= AFFINITY_FLOOR)
    // Strongest opinions first — in EITHER direction. A firm dislike is as
    // useful to the ranker as a firm like, and truncating by signed value would
    // throw away every dislike before the first weak like.
    .sort((a, b) => Math.abs(b[1]) - Math.abs(a[1]) || a[0].localeCompare(b[0]))
    .slice(0, MAX_AFFINITIES);
  for (const [item, value] of scored) affinities[item] = value;

  return { affinities, interactions };
}

File: apps/recipe-service/internal/recommend/affinity.go
package recommend

// --- Ingredient affinity (BL-0005 increment 2) ---------------------------
//
// `affinity` was wired and permanently unavailable in increment 1: it needs an
// interaction log, and there was none. There is one now, and this file is the
// whole of what the ranker knows about it — which is deliberately very little.
//
// The RANKER NEVER SEES AN EVENT. Convex owns `recommendationEvents`, folds a
// recent window of them into an ingredient→weight map, and sends that map in the
// request. This package stays dependency-free and stateless, and the derivation
// stays in the one place that owns the history it derives from. Nothing is
// persisted as a score, so there is no derived table to drift from the events
// that produced it.
//
// Two rules here are load-bearing:
//
//   - **A cold-start user makes affinity UNAVAILABLE, never zero.** A user with
//     no history and no stated likes has told us nothing about their taste.
//     Scoring that as 0 would put every candidate at the bottom of a feature
//     that carries real weight, punishing precisely the users who have not used
//     the product yet. `available: false` removes it from both sides of
//     combine()'s average instead, which leaves the remaining features scoring
//     exactly as they did before this file existed.
//   - **Availability is REQUEST-LEVEL, not per candidate.** Deciding it per
//     candidate would normalize two rows of one response by different weight
//     denominators, making their scores incomparable — the exact failure
//     combine() exists to prevent (see pantryView.hasExpiry, which learned this
//     first). So a candidate containing nothing we have an opinion about scores
//     0, which is a real observation ("nothing here is something you've reacted
//     to"), not a missing one.

// affinitySaturation is how many opinionated ingredients a recipe needs before
// it can reach a full ±1 on this feature.
//
// Without it a recipe whose single matched ingredient is garlic would score the
// same as one built entirely out of things the user loves, because a bare mean
// over matched items is blind to how much agreement it is averaging. Dividing by
// at least this many says what we mean: more of the recipe pointing the same way
// is a stronger signal than one line pointing very hard.
const affinitySaturation = 3

// explicitWeight is what a stated like or dislike is worth on the same scale the
// derived weights live on. It is the maximum: the user telling us directly is
// the strongest evidence about their taste that exists.
const explicitWeight = 1.0

// affinityView is the request's ingredient opinions, built once per request
// rather than per candidate.
type affinityView struct {
	// scores maps canonicalItem → [-1, 1]. Absent means no opinion, which is
	// distinct from an opinion of zero and is why this is a map rather than a
	// function returning 0.
	scores map[string]float64
	// available is the cold-start switch, and it is request-level. See the
	// header: per-candidate availability would make rows incomparable.
	available bool
}

// newAffinityView merges the two sources of ingredient opinion the ranker has.
//
// The DERIVED map arrives from Convex's fold over the event log. The EXPLICIT
// lists arrive from the user's own settings, and they OVERRIDE the derived value
// outright rather than averaging with it. Averaging would let a month of
// inferred behaviour argue with something the user typed on purpose, and between
// "we noticed" and "they said", they said wins. (Explicit likes are also why a
// brand-new user who filled in the settings form gets a live affinity feature
// without having interacted with a single recommendation.)
func newAffinityView(uc UserContext) affinityView {
	scores := make(map[string]float64, len(uc.Affinities)+
		len(uc.Preferences.LikedItems)+len(uc.Preferences.DislikedItems))

	for item, weight := range uc.Affinities {
		if item == "" {
			continue
		}
		scores[item] = clampSigned(weight)
	}
	// Dislikes are applied after likes so that an item on both lists resolves to
	// the more cautious answer, the same way the avoid list resolves ties by
	// removing rather than by keeping.
	for _, item := range uc.Preferences.LikedItems {
		if item != "" {
			scores[item] = explicitWeight
		}
	}
	for _, item := range uc.Preferences.DislikedItems {
		if item != "" {
			scores[item] = -explicitWeight
		}
	}

	return affinityView{scores: scores, available: len(scores) > 0}
}

// score reports how much this candidate looks like things the user likes, in
// [-1, 1]. Only ingredients we have an OPINION about contribute; the rest are
// neither evidence for nor against.
//
// De-duplicated by canonical item for the same reason matchCandidate is: a
// recipe listing garlic in two units is not twice as garlicky.
func (v affinityView) score(c Candidate) float64 {
	if !v.available {
		return 0
	}
	var sum float64
	var matched int
	seen := make(map[string]bool, len(c.Ingredients))
	for _, ing := range c.Ingredients {
		if seen[ing.CanonicalItem] {
			continue
		}
		seen[ing.CanonicalItem] = true
		w, ok := v.scores[ing.CanonicalItem]
		if !ok {
			continue
		}
		sum += w
		matched++
	}
	if matched == 0 {
		return 0
	}
	denom := matched
	if denom < affinitySaturation {
		denom = affinitySaturation
	}
	return clampSigned(sum / float64(denom))
}

// affinityFeature is shared by BOTH surfaces. Discovery weights it heavily and
// the pantry surface lightly, but the value and its availability are computed
// identically — two definitions of "you like this" would eventually disagree,
// and the user would have no way to tell which screen was lying.
func affinityFeature(value float64, v affinityView, w Weights) feature {
	return feature{
		name:      "affinity",
		value:     value,
		weight:    w.Affinity,
		available: v.available,
	}
}

// affinityReasonThreshold is how strong the signal has to be before the card
// claims it out loud. A faint positive still ranks; it just does not boast.
const affinityReasonThreshold = 0.34

// affinityReason explains the feature, and only ever in the positive direction.
// "Built on things you don't like" is a true statement we have no business
// putting on screen — the recipe is already ranked below the alternatives, which
// is the entire action a negative signal justifies.
func affinityReason(value float64) string {
	if value < affinityReasonThreshold {
		return ""
	}
	return "Uses things you cook a lot"
}

func clampSigned(x float64) float64 {
	if x < -1 {
		return -1
	}
	if x > 1 {
		return 1
	}
	return x
}
